### PR TITLE
refactor cppmanglewin to use slices, extern(D) and identifiers

### PR DESCRIPTION
**Pull request by thewilsonator**
make private function `extern(D)` as well.

### CODE REVIEW

File: src/dmd/cppmanglewin.d [hunk 1]
```diff
@@ -897,10 +897,33 @@ private:
...
+                }
+            }
+            buf.writestring(name);
+            buf.writeByte('@');
+        }
         if (TemplateInstance ti = sym.isTemplateInstance())
         {
             auto id = ti.tempdecl.ident;
-            const(char)* symName = id.toChars();
+            auto symName = id.toString();
```
Comment (WalterBright) [inline src/dmd/cppmanglewin.d:926]: It appears that `symName` is *always* an `Identifier` and `saved_idents` are *always* `Identifiers`. If true, all of these strings can be replaced with `Identifier` and then one never has to use the string comparisons.

File: src/dmd/cppmanglewin.d [hunk 2]
```diff
@@ -997,7 +1000,7 @@ private:
                 saved_idents[i] = name;
                 break;
             }
-            if (!strcmp(saved_idents[i], name)) // ok, we've found same name. use index instead of name
+            if (saved_idents[i] == name) // ok, we've found same name. use index instead of name
```
Comment (WalterBright) [inline src/dmd/cppmanglewin.d:1003]: A case where the string comparison can be eliminated if it is switched to `Identifier`s instead.

File: src/dmd/cppmanglewin.d (hunk 1 above)
Comment (thewilsonator) [reply src/dmd/cppmanglewin.d:926]: I'm doing this reactor for https://github.com/dlang/dmd/pull/11835 in which mangle overrides are strings not idents.

Comment (thewilsonator) [reply src/dmd/cppmanglewin.d:926]: Actually this can be made to work.

File: src/dmd/cppmanglewin.d [hunk 3]
```diff
@@ -958,37 +982,17 @@ private:
                     fatal();
                 }
             }
-            name = tmp.buf.extractChars();
+            writeName(Identifier.idPool(tmp.buf.extractSlice()));
```
Comment (thewilsonator) [inline src/dmd/cppmanglewin.d:985]: @WalterBright this is the only price that we have to pay to use identifiers.

File: src/dmd/cppmanglewin.d [hunk 4]
```diff
@@ -897,10 +898,33 @@ private:
             return;
         }
 
+        void writeName(Identifier name)
+        {
+            assert(name);
+            if (is_dmc_template)
+            {
+                if (checkAndSaveIdent(name))
+                return;
```
Comment (Geod24) [inline src/dmd/cppmanglewin.d:907]: Indent

File: src/dmd/cppmanglewin.d [hunk 5]
```diff
@@ -897,10 +898,33 @@ private:
...
+            else
+            {
+                if (dont_use_back_reference)
+                {
+                    saveIdent(name);
+                }
+                else
+                {
+                    if (checkAndSaveIdent(name))
+                    return;
```
Comment (Geod24) [inline src/dmd/cppmanglewin.d:918]:
```suggestion
                else if (checkAndSaveIdent(name))
                    return;
```

File: src/dmd/cppmanglewin.d [hunk 6]
```diff
@@ -897,10 +898,33 @@ private:
...
+            {
+                if (checkAndSaveIdent(name))
+                return;
+            }
+            else
+            {
+                if (dont_use_back_reference)
+                {
+                    saveIdent(name);
+                }
```
Comment (Geod24) [inline src/dmd/cppmanglewin.d:914]: Couldn't this function be simplified ?
```D
if (!is_dmc_template && dont_use_back_reference)
    saveIdent(name)
else if (checkAndSaveIdent(name))
    return;
```

?

**Review by Geod24 — APPROVED** [review]
Nice, just a few nits

**Pull request merged**